--- a/src/http/ngx_http_core_module.c	2025-02-25 00:44:29.376263119 +0000
+++ b/src/http/ngx_http_core_module.c	2025-02-25 00:47:38.804263119 +0000
@@ -1959,8 +1959,13 @@ ngx_http_auth_basic_user(ngx_http_reques
     ngx_str_t   auth, encoded;
     ngx_uint_t  len;
 
-    if (r->headers_in.user.len == 0 && r->headers_in.user.data != NULL) {
-        return NGX_DECLINED;
+    if (r->headers_in.user.data != NULL) {
+        if (r->headers_in.user.len == 0) {
+	    return NGX_DECLINED;
+	} else {
+	    /* headers_in.user is already set, don't parse again. */
+	    return NGX_OK;
+	}
     }
 
     if (r->headers_in.authorization == NULL) {
